src/Storages/MergeTree/MergeTreeIndexHnsw.cpp

```diff
@@ -0,0 +1,351 @@
+#ifdef ENABLE_USEARCH
+
+#include <Storages/MergeTree/MergeTreeIndexHnsw.h>
+
+#include <Columns/ColumnArray.h>
+#include <Common/typeid_cast.h>
+#include <Core/Field.h>
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/DataTypeTuple.h>
+#include <IO/ReadHelpers.h>
+#include <IO/WriteHelpers.h>
+#include <Interpreters/Context.h>
+#include <Interpreters/castColumn.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_COLUMN;
+    extern const int INCORRECT_DATA;
+    extern const int INCORRECT_NUMBER_OF_COLUMNS;
+    extern const int INCORRECT_QUERY;
+    extern const int LOGICAL_ERROR;
+}
+
+
+template <unum::usearch::metric_kind_t Metric>
+USearchIndexWithSerialization<Metric>::USearchIndexWithSerialization(size_t dimensions)
+    : Base(Base::make(unum::usearch::metric_punned_t(dimensions, Metric)))
+{
+}
+
+template <unum::usearch::metric_kind_t Metric>
+void USearchIndexWithSerialization<Metric>::serialize([[maybe_unused]] WriteBuffer & ostr) const
+{
+    auto callback = [&ostr](void * from, size_t n)
+    {
+        ostr.write(reinterpret_cast<const char *>(from), n);
+        return true;
+    };
+
+    Base::stream(callback);
+}
+
+template <unum::usearch::metric_kind_t Metric>
+void USearchIndexWithSerialization<Metric>::deserialize([[maybe_unused]] ReadBuffer & istr)
+{
+    BufferBase::Position & pos = istr.position();
+    unum::usearch::memory_mapped_file_t memory_map(pos, istr.buffer().size() - istr.count());
+    Base::view(std::move(memory_map));
+    pos += Base::stream_length();
+
+    auto copy = Base::copy();
+    if (!copy)
+        throw std::runtime_error("Can't copy index");
```

Comment:
ClickHouse has its own Exceptions, let's rewrite to something like
```cpp
throw Exception(ErrorCodes::LOGICAL_ERROR, "Could not copy usearch index");
```

Reply:
Alternatively, we can implement a different form of deserialization, but memory-mapping fragments of the buffer and copying afterward at once is simpler. In this case, swapping is identical to move-assignment. We can refactor to the latter, if you prefer.

Reply:
As for exceptions, they should also be handled in `add` and `search`. @davvard will adjust the code accordingly.